src/test/java/com/google/sps/ProfileBuilderServletTest.java

```diff
@@ -0,0 +1,196 @@
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.gson.Gson;
+import com.google.sps.data.User;
+import com.google.sps.data.PersistentUserRepository;
+import com.google.sps.servlets.ProfileBuilderServlet;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+public class ProfileBuilderServletTest {
+    
+  @Mock
+  private HttpServletRequest request;
+
+  @Mock
+  private HttpServletResponse response;
+
+
+  private ProfileBuilderServlet userServlet;
+
+  private LocalServiceTestHelper localHelper =
+    new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+    
+    
+  @Before
+  public void setUp() throws Exception {
+    userServlet = new ProfileBuilderServlet(); 
+    request = Mockito.mock(HttpServletRequest.class);       
+    response = Mockito.mock(HttpServletResponse.class);
+
+    localHelper.setUp();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    localHelper.tearDown();
+  }
+    
+  public void addMentorProfile() {
+    String userType = "Mentor";
+    when(request.getParameter("user-Type")).thenReturn(userType);
+    String mentorName = "Jeff";
+    when(request.getParameter("profName-input")).thenReturn(mentorName);
+    String company = "Google";
+    when(request.getParameter("company-input")).thenReturn(company);
+    String specialty = "Security";
+    when(request.getParameter("specialty-input")).thenReturn(specialty);
+  }
+
+  public void addMenteeProfile() {
+    String userType = "Mentee";
+    when(request.getParameter("user-Type")).thenReturn(userType);
+    String menteeName = "Bobby";
+    when(request.getParameter("name-input")).thenReturn(menteeName);
+    String school = "University of California Berkeley";
+    when(request.getParameter("school-input")).thenReturn(school);
+    String major = "Computer Science";
+    when(request.getParameter("major-input")).thenReturn(major);
+  }
+
+  @Test
+  public void doPostRedirectsUrl() throws IOException, ServletException {
+    addMentorProfile();
```

Comment:
Is there a reason we need to call addMentorProfile() for testing that a POST request will send a redirect?

Reply:
I just call it to add the fake user data so that the post request posts the user information to datastore. Without it the post method cannot properly function and I get a null pointer exception.

Reply:
The test pretty much checks to see that the doPost posts the user information to datastore then redirects at the end.

Reply:
I read that this test verifies that posting user information to the servlet causes a redirect. 

Could you also include a test where a POST request has missing user information? We should ensure that the empty request case is well tested to prevent the servlet from throwing a null pointer exception in production.

Reply:
For this additional test, would doing this check be sufficient

1. Add mentor profile with a missing information : specialty in this case
2. call Post request
3. do this assert test 
`    Assert.assertEquals(1, dataService.prepare(new Query("User")).countEntities(withLimit(10)));
`

This test returns true.

Reply:
Testing with one empty field field (e.g. specialty) provides some coverage. I am more concerned about the case where the user makes a POST request without including _any_ fields. You can write this test using the following guideline:

1.  Create a mock request where all accessed fields return a nullptr.
2. Call the POST method with the mock request.
3. Verify that the servlet returns a redirect and does not store any new data into datastore.

Reply:
For step 3, is that 2 checks

like the verify it redirects

and the assert equals

Reply:
Also, for step 1
shoultd it return null or nullptr

Nullptr said no symbol found


Then with null that just led to a nullptr exception error

Reply:
Sorry for the confusion, please use null instead of nullptr. If there is a nullptr exception error, your servlet may not be handling this case correctly.

Reply:
Ok

Reply:
Done